pkg/kvcache/metrics/collector.go

```diff
@@ -0,0 +1,118 @@
+package metrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+var (
+	Admissions = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "kvcache", Subsystem: "index", Name: "admissions_total",
+		Help: "Total number of KV-block admissions",
+	})
+	Evictions = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "kvcache", Subsystem: "index", Name: "evictions_total",
+		Help: "Total number of KV-block evictions",
+	})
+
+	// LookupRequests counts how many Lookup() calls have been made.
+	LookupRequests = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "kvcache", Subsystem: "index", Name: "lookup_requests_total",
+		Help: "Total number of lookup calls",
+	})
+	// KeyLookupResults counts keys examined, labelled by result “hit” or
+	// “miss”.
+	KeyLookupResults = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "kvcache", Subsystem: "index", Name: "lookup_key_total",
+		Help: "Number of keys looked up by result (hit/miss)",
+	}, []string{"result"})
```

Comment:
Trying to decipher what these mean.
if you have two counters, one for total lookups and one for total hits, you can use PromQL later to calculate rates over time periods as well as "trends".
The `LookupRequests` seems to be the total lookups. Is `KeyLookupResults` just two counters - one for hit and one for miss?
Can't you do miss-ratio = (total-hit)/total on the fly instead of counting both?

Reply:
Makes sense, done.